compiler/packages/babel-plugin-react-compiler/src/Optimization/ConstantPropagation.ts

```diff
@@ -472,6 +472,52 @@ function evaluateInstruction(
       }
       return null;
     }
+    case "TemplateLiteral": {
+      if (value.quasis.some((q) => q.cooked === undefined)) {
+        return null;
+      }
+
+      if (value.subexprs.length === 0) {
+        const result: InstructionValue = {
+          kind: "Primitive",
+          value: value.quasis.map((q) => q.cooked).join(""),
+          loc: value.loc,
+        };
+        instr.value = result;
+        return result;
+      }
+
+      const subExpressionValues = value.subexprs.map((subExpr) =>
+        read(constants, subExpr)
+      );
+
+      let resultString = value.quasis[0].cooked;
+      let quasiIndex = 1;
+      for (const subExprValue of subExpressionValues) {
+        if (!subExprValue) {
+          return null;
+        }
+        if (subExprValue.kind !== "Primitive") {
+          return null;
+        }
+        if (subExprValue.value === null || subExprValue.value === undefined) {
+          return null;
+        }
+
+        const suffix = value.quasis[quasiIndex].cooked;
+        resultString += subExprValue.value.toString() + suffix;
```

Comment:
can you double-check the spec behavior here? is it calling toString() or something else? would be great to link to the appropriate point in the spec for posterity.

Reply:
This is a good concern. I've looked it up:
https://262.ecma-international.org/6.0/#sec-template-literals-runtime-semantics-evaluation

It uses string coercion (defined as ToString), wich is defined here: https://262.ecma-international.org/6.0/#sec-tostring
We only deal with string/number/boolean in the case above. So I think the only case that would be non-trivial is `ToString(number)`, which is defined [here](https://262.ecma-international.org/6.0/#sec-tostring-applied-to-the-number-type). [This](https://262.ecma-international.org/6.0/#sec-number.prototype.tostring) is the definition of `Number.prototype.toString()`, which - in case of radix = 10 - just returns `ToString(number)`.

If I'm not mistaking, it matches the spec as long as the values are number/string/booleans. It should be a different story when it is null/undefined (which is excluded via the return above).

In the note it says that the coercion basically behaves like when using [String.prototype.concat](https://262.ecma-international.org/6.0/#sec-string.prototype.concat), which is also the way the TS compiler implements this when lowering to ES3/ES5 (they seem to assume that concat operation is not monkey-patched).

Is this fine or should we just handle the cases where it's trivially always a string? More sophisticated stuff can always be added later.